api/game/clueless.py

```diff
@@ -24,56 +29,226 @@ def __init__(self, *args, **kwargs):
+        )
+        to_be_distributed = list(to_be_distributed)
+        random.shuffle(to_be_distributed)
+        
+        # determine how many cards each player gets
+        n_cards = len(to_be_distributed)
+        n_players = len(self.players)
+        n = n_cards // n_players
+
+        # distribute 
+        split = [to_be_distributed[i:i + n] for i in range(0, n_cards, n)]
+        player_cards = dict(zip(self.players.keys(), split[:-1]))
+        visible_cards = split[-1]
+
+        return player_cards, visible_cards
+
+
+    def generate_turn_order(self) -> List:
+        """ Clue rules state Miss scarlet moves first, and then play proceeds
+        clockwise """
+
+        # if miss scarlet isn't a chosen player token
+        player_tokens = list(self.players.values())
+
+        token_order = [
+            'miss_scarlet',
+            'colonel_mustard',
+            'mrs_white',
+            'mr_green',
+            'mrs_peacock',
+            'professor_plum',
+        ]
+        for token in token_order:
+            if token in player_tokens:
+                break
+        turn_order = [token]
+
+        # no concept of clockwise in UI currently, so randomize for now
+        random.shuffle(player_tokens)
+        for player_token in player_tokens:
+            if player_token != token: # first player already chosen
+                turn_order.append(player_token)
+
+        assert len(player_tokens) == len(turn_order) # future unit test 
+
+        return turn_order
+
+
+    def get_next_player(self, player: str) -> Tuple:
+        """ Returns the next token to move and their client ID. 
+        Note: does not update the game state, expected to be handled by caller.
+        This function is also used to determine next player up to disprove a 
+        suggestion.
+
+        Args:
+            player: current player's token (i.e. 'professor_plum')
+        """
+
+        idx = self.state['turn_order'].index(player)
+        
+        # get next player token
+        try:
+            next_player = self.state['turn_order'][idx + 1]
+        except IndexError:
+            next_player = self.state['turn_order'][0]
+        
+        # get client id associated with next player
+        for client_id, token in self.players.items():
+            if next_player == token:
+                break
+        
+        return next_player, client_id
+
+
+    def allowed_moves(self, player: str) -> List:
+        """ Determines locations player may move their token
+
+        Args:
+            player: current player's token (i.e. 'professor_plum')
+        
+        Returns:
+            A list of allowed locations
+        """
+        allowed_moves = []
+        
+        # get current token location
+        current_loc = self.state['suspect_locations'][player]
+        
+        # if first time through turn rotation, can only move into a hallway
+        # TODO track first turn or pass as boolean into this function

```

Comment:
Could we make the starting locations an attribute of this class like self.hallways and self.rooms? I think we should treat them like rooms except you can't enter them.

Reply:
done